src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentNavigationService.cs

```diff
@@ -292,18 +298,18 @@ private bool TryNavigateToMappedFile(Workspace workspace, Document generatedDocu
                     iEndLine = mappedSpanResult.LinePositionSpan.End.Line
                 };
 
-                return NavigateTo(textBuffer, vsTextSpan);
+                return NavigateTo(textBuffer, vsTextSpan, cancellationToken);
             }
 
             return false;
         }
 
-        private static MappedSpanResult? GetMappedSpan(ISpanMappingService spanMappingService, Document generatedDocument, TextSpan textSpan)
+        private static MappedSpanResult? GetMappedSpan(ISpanMappingService spanMappingService, Document generatedDocument, TextSpan textSpan, CancellationToken cancellationToken)
```

Comment:
yup.  seems like we want this to be async.

Reply:
@CyrusNajmabadi there are quite a few places that this gets called from synchronous editor APIs.  should we just make those task/jtf.run?  In a couple of those places they also don't pass a cancellation token

Reply:
i owuldn't JTF run.  i would make a cancellable TWD and just kick off the async work.  when the async work finished, we tear down the TWD.

this way the user can actually know what's happening, and can cancel out if they want.

Reply:
I see, in many cases that should be possible as they're invoked from a command handler which passes the CommandExecutionContext which has the IUIThreadOperationContext

However for things like [INavigableSymbol](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FLanguage%2FDef%2FIntellisense%2FNavigableSymbol%2FINavigableSymbol.cs&version=GBmain&line=27&lineEnd=28&lineStartColumn=1&lineEndColumn=1&lineStyle=plain) there isn't a context passed along to us, so I'm not exactly sure how to popup the dialog.  @olegtk is it possible to create our own dialog here, or otherwise update the APIs to pass in a context we can use?

Reply:
Those Navigate() calls are basically fire and forget to the editor, we don't need any results back, we just fully delegate navigation to you. INavigableSymbol.Navigate() is such case, there is no existing TWD or commanding context to pass, so I'd suggest you create you own TWD for such scenarios.

Reply:
Yeah, we should not make our own threaded wait dialog *here*, but in all the other places (many of which I called out in this review) make one and then pass that cancellation token through.

Whether we want to make this async....I'm not sure I see the value given a lot of the "interesting" work here is still synchronous like the call to OpenDocumentFromPath; sure you could make it async but I don't think any user experience actually _changes_, so I'm guessing we probably have better work to do? Now if we had a bunch of other stuff to also consume async APIs (like async file open that @olegtk was working on) then that might have a much nicer value. But in that case, a better experience might be not to have a modal threaded wait dialog up but instead just have the tab open, and let the editor move to our position when it completes the async file load. Otherwise we're putting async atop async.